[4기 고예성] Mission 1 PR 제출합니다.

Dev-Yesung opened the pull request:
## 📌 과제 설명
- [x]  JPA 프로젝트를 세팅해본다.
- [x]  세팅한 프로젝트를 이용해서 단일 엔티티를 이용한 CRUD를 구현한다.
    - 고객(Customer) 엔티티는 ID(PK), 이름, 성을 가진다.
    - 고객엔티티를 이용한 CRUD를 구현한다.

Dev-Yesung committed:
feat: Customer 엔티티 생성

JPA에 매핑될 Customer 엔티티를 생성하였습니다.

Dev-Yesung committed:
feat: Customer CRUD 테스트 작성

CRUD 테스트를 작성했습니다.

kimihiqq commented on src/main/java/kr/co/prgrms/jpaintro/domain/Customer.java:
```diff
@@ -0,0 +1,71 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
+    private Long id;
+    @Column(name = "first_name", nullable = false, length = 50)
+    private String firstName;
+    @Column(name = "last_name", nullable = false, length = 50)
+    private String lastName;
```
`length = 50`를 활용하여 데이터 필드에 제한을 설정해주신 부분 좋은 것 같습니다! 👍👍

Dev-Yesung replied:
ㅎㅎㅎ 감사합니다~ JPA가 처음이라 낯설어서 사용할 수 있는 기능들을 최대한 활용해보려고 했습니다!

kimihiqq commented on src/main/java/kr/co/prgrms/jpaintro/domain/Customer.java:
```diff
@@ -0,0 +1,71 @@
+    protected Customer() {
+    }
+
+    public Customer(String firstName, String lastName) {
+        this.firstName = firstName;
+        this.lastName = lastName;
+    }
+
+    public Customer(Long id, String firstName, String lastName) {
+        validateName(firstName);
+        validateName(lastName);
+        this.id = id;
+        this.firstName = firstName;
+        this.lastName = lastName;
+    }
```
현재 생성자가 두개인데요, `@GeneratedValue(strategy = GenerationType.SEQUENCE)`를 통해 id가 자동으로 생성되니 id를 매개변수로 받는 생성자는 없어도 되지 않을까..? 생각합니다!

Dev-Yesung replied:
제가 정확하게 몰라서 그런데, JPA는 DB에서 데이터를 가져올 때, 생성자를 가지고 가져오지 않는건가요...? 
혹시나 해서 DB에서 데이터를 가져올 때 생성자를 통해 가져온다면 id에 관한 생성자도 필요할까봐 이렇게 만들었습니다...! 
이 부분 다시 공부하고 수정하도록 하겠습니다 ㅎㅎ

이와 별개로 제가 이름만 가지고 생성하는 생성자에는 validName을 안 걸어놨네요... ㅜ🥲
수정하도록 하겠습니다! 감사합니다~

Dev-Yesung replied:
이 정보가 맞는지 모르겠는데 내부적인 메커니즘상 기본생성자를 만드는게 제일 중요하군요...! 수정하겠습니다~

https://chat.openai.com/share/a91172f5-5778-43d1-9c66-8321720753be

kimihiqq replied:
JPA가 DB에서 데이터를 가져올때는 기본적으로 (매개변수가 없는)기본생성자를 사용하는 것으로 알고있습니다..! 이후 Reflection API를 사용하여 private 필드에 직접 접근하고 값을 변경하기에, id를 매개변수로 받는 생성자도 사용하지 않는것으로 저는 알고있습니다..!

다음 글을 참고해보셔도 좋을 것 같습니다!
https://velog.io/@yyy96/JPA-%EA%B8%B0%EB%B3%B8%EC%83%9D%EC%84%B1%EC%9E%90

Dev-Yesung replied:
오 저도 이 글 읽었는데, 결국 기본생성자가 반드시 필요한 이유와 동일하군요!! 이해됐습니다 ㅎㅎ 자세한 답변 감사합니다~

kimihiqq commented on src/main/java/kr/co/prgrms/jpaintro/domain/Customer.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    private void checkIllegalCharacter(String name) {
+        String pattern = "^[a-zA-Zㄱ-ㅎ가-힣]*$";
+
+        if (!Pattern.matches(pattern, name)) {
+            throw new IllegalNameException("[ERROR] 이름 값은 한글과 영문만 가능합니다!");
+        }
+    }
```
👍👍 유효성 검사를 통해 이름이 특정 패턴에 일치하는지 확인해주셨는데, 이부분은 저도 참고해 추가하겠습니다ㅎㅎ~

Dev-Yesung replied:
저도 공부하다가 어디서 참고해서 갖다 썼습니다 ㅎㅎ

kimihiqq commented on src/main/java/kr/co/prgrms/jpaintro/domain/Customer.java:
```diff
@@ -0,0 +1,71 @@
+    @Column(name = "last_name", nullable = false, length = 50)
+    private String lastName;
+
+    protected Customer() {
+    }
```
이부분은 lombok의 @NoArgsConstructor의 사용을 고려해볼수도 있을 것 같습니다!

Dev-Yesung replied:
아앗 저는 롬복에서 제공하는 어노테이션에 public으로만 가능한줄 알았는데 
protected로 설정하는 것도 가능하군요...!!!🤩
@NoArgConstructor를 사용해서 바꿔보도록 하겠습니다~ 감사합니다 ㅎㅎ

kimihiqq reviewed (approved):
깔끔하게 코드를 잘 작성해주셔서 리뷰드릴부분이 많이 없어서.. 좋았던 점 비롯해서 몇가지 남겨 보았습니다ㅎㅎ!

예성님 수고많으셨습니다~👍👍

Dev-Yesung committed:
refactor: 코드리뷰 피드백사항 반영

@GeneratedValue를 통해 자동으로 id가 생성되기 때문에 매개변수가 필요없어 생성자를
수정했습니다.
기본생성자가
protected로 선언된 것에서 @NoArgConstructor(access = AccessLevel.PROTECTED)로
변경했습니다.
